### PR TITLE
[Perf] Eliminate padding and slicing op for GPT-OSS with Flashinfer MXFP4 MXFP8 MoE

### PR DESCRIPTION
**Pull request by elvischenv**
## Purpose

- Depends on Flashinfer update #30993
- Eliminated padding op before the MoE: by setting the alignment in flashinfer mxfp8 quant, the output quantized tensor will be padded.
- Eliminated slicing op after the MoE: by passing the output tensor with unpadded hidden size to MoE kernel, this depends on a Flashinfer PR:
  - https://github.com/flashinfer-ai/flashinfer/pull/2217
  - This will also resolve the previous AR+Norm fusion broken by slice op issue: #28841
- ~~Cleaned up the padding logic: for mxfp4 quant, the padded hidden size is calculated in `create_weights()`, the `maybe_roundup_hidden_size()` in `vllm/model_executor/layers/fused_moe/layer.py` seems like a dup.~~

### Accuracy

PR:
```
[{'eval_name': 'gpqa', 'model_name': 'gpt-oss-120b-high_temp1.0_20251213_233136', 'metric': 0.7803030303030303}]
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-high_temp1.0_20251213_234320', 'metric': 0.8875}]
```

main:
```
[{'eval_name': 'gpqa', 'model_name': 'gpt-oss-120b-high_temp1.0_20251214_002509', 'metric': 0.7891414141414141}]
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-high_temp1.0_20251214_001505', 'metric': 0.8875}]
```

### Kernel

PR:
```
void cublasLt::splitKreduce_kernel                          2.400 μs
void tensorrt_llm::kernels::quantize_with_block_size        2.944 μs
void moe::dev::routing::routingRenormalize                  5.216 μs
bmm_MxE4m3_MxE2m1MxE4m3_Fp32_t128x16x256u2                 13.216 μs
bmm_Bfloat16_MxE2m1MxE4m3_Fp32_t128x16x256u2                9.504 μs
void moe::dev::finalize::finalizeKernel                     3.168 μs
void flashinfer::trtllm_allreduce_fusion                    7.744 μs (ar+norm)
nvjet_tst_32x64_64x16_4x1_v_bz_splitK_TNN                   4.448 μs
```

main:
```
void cublasLt::splitKreduce_kernel                          2.048 μs
triton_poi_fused_constant_pad_nd_moe_forward_0              1.407 μs (pad)
void tensorrt_llm::kernels::quantize_with_block_size        2.432 μs
void moe::dev::routing::routingRenormalize                  5.056 μs
bmm_MxE4m3_MxE2m1MxE4m3_Fp32_t128x16x256u2                 10.368 μs
bmm_Bfloat16_MxE2m1MxE4m3_Fp32_t128x16x256u2                8.512 μs
void moe::dev::finalize::finalizeKernel                     2.112 μs
void vllm::cross_device_reduce_1stage                       8.320 μs (ar)
triton_red_fused__to_copy_add_mean_mul_pow_rsqrt_slice_1    2.336 μs (norm, slice)
nvjet_tst_32x64_64x16_4x1_v_bz_splitK_TNN                   4.448 μs
```

### Perf (GPT-OSS-120b TP8 con8)

PR: **5% E2E improvement**
```
============ Serving Benchmark Result ============
Successful requests:                     80
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  28.90
Total input tokens:                      81920
Total generated tokens:                  81920
Request throughput (req/s):              2.77
Output token throughput (tok/s):         2834.13
Peak output token throughput (tok/s):    158.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          5668.25
---------------Time to First Token----------------
Mean TTFT (ms):                          50.37
Median TTFT (ms):                        58.07
P99 TTFT (ms):                           75.11
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.77
Median TPOT (ms):                        2.77
P99 TPOT (ms):                           2.84
---------------Inter-token Latency----------------
Mean ITL (ms):                           54.59
Median ITL (ms):                         55.28
P99 ITL (ms):                            68.89
==================================================
```

main:
```
============ Serving Benchmark Result ============
Successful requests:                     80
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  30.43
Total input tokens:                      81920
Total generated tokens:                  81920
Request throughput (req/s):              2.63
Output token throughput (tok/s):         2692.11
Peak output token throughput (tok/s):    149.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          5384.22
---------------Time to First Token----------------
Mean TTFT (ms):                          52.61
Median TTFT (ms):                        57.20
P99 TTFT (ms):                           88.23
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.92
Median TPOT (ms):                        2.92
P99 TPOT (ms):                           2.99
---------------Inter-token Latency----------------
Mean ITL (ms):                           57.47
Median ITL (ms):                         58.17
P99 ITL (ms):                            70.35
==================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: tests/compile/fusions_e2e/conftest.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ def run(
         tp_size: int = 1,
     ):
         monkeypatch.setenv("VLLM_USE_DEEP_GEMM", "1" if use_deepgemm else "0")
+        monkeypatch.setenv("VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8", "1")
```
Comment (elvischenv) [inline tests/compile/fusions_e2e/conftest.py]: @ProExpertProg Wondering a better way to add env for a specific model

Comment (ProExpertProg) [reply tests/compile/fusions_e2e/conftest.py]: `KernelConfig` was just merged recently, you can add to that!

Comment (elvischenv) [reply tests/compile/fusions_e2e/conftest.py]: Do you mean we want to deprecate the kernel envs(e.g. `VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8`), and migrate to KernelConfig(e.g. `use_flashinfer_moe_mxfp4_mxfp8`) and flag(e.g. `--use-flashinfer-moe-mxfp4-mxfp8`)?

Comment (ProExpertProg) [reply tests/compile/fusions_e2e/conftest.py]: Yes! Although to nitpick I don't think this deserves its own top-level flag.

If this is an existing env car it can stay like that and be migrated later also

Comment (ProExpertProg) [reply tests/compile/fusions_e2e/conftest.py]: But yeah for E2E tests this is fine for now. Later once migrated to kernel config, we can just add it to model_kwargs!

Comment (ProExpertProg) [reply tests/compile/fusions_e2e/conftest.py]: If you want this only applied to gpt-oss, could always if check the model name. And add a comment with a TODO mentioning the migration to model kwargs

Comment (elvischenv) [reply tests/compile/fusions_e2e/conftest.py]: > If you want this only applied to gpt-oss, could always if check the model name. And add a comment with a TODO mentioning the migration to model kwargs

Updated in this way.

**Comment by elvischenv** [comment]
> @elvischenv could you post the new benchmarking numbers once you have them?

This is the perf number based on main ToT:
PR
```
============ Serving Benchmark Result ============
Successful requests:                     40
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  15.67
Total input tokens:                      40960
Total generated tokens:                  40960
Request throughput (req/s):              2.55
Output token throughput (tok/s):         2613.70
Peak output token throughput (tok/s):    141.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          5227.39
---------------Time to First Token----------------
Mean TTFT (ms):                          61.13
Median TTFT (ms):                        52.89
P99 TTFT (ms):                           93.71
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          3.00
Median TPOT (ms):                        3.00
P99 TPOT (ms):                           3.05
---------------Inter-token Latency----------------
Mean ITL (ms):                           59.05
Median ITL (ms):                         59.72
P99 ITL (ms):                            83.17
==================================================
```

main
```
============ Serving Benchmark Result ============
Successful requests:                     40
Failed requests:                         0
Maximum request concurrency:             8
Benchmark duration (s):                  16.46
Total input tokens:                      40960
Total generated tokens:                  40960
Request throughput (req/s):              2.43
Output token throughput (tok/s):         2488.50
Peak output token throughput (tok/s):    136.00
Peak concurrent requests:                16.00
Total token throughput (tok/s):          4977.00
---------------Time to First Token----------------
Mean TTFT (ms):                          64.64
Median TTFT (ms):                        64.21
P99 TTFT (ms):                           101.20
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          3.15
Median TPOT (ms):                        3.15
P99 TPOT (ms):                           3.22
---------------Inter-token Latency----------------
Mean ITL (ms):                           62.04
Median ITL (ms):                         63.02
P99 ITL (ms):                            64.60
==================================================
```

accuracy of PR
```
[{'eval_name': 'gpqa', 'model_name': 'gpt-oss-120b-high_temp1.0_20260208_183703', 'metric': 0.7992424242424242}]
```

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -221,61 +224,6 @@ def get_compressed_expert_map(expert_map: torch.Tensor) -> str:
...
-        if (
-            current_mxfp4_backend == Mxfp4Backend.SM90_FI_MXFP4_BF16
-            or current_mxfp4_backend == Mxfp4Backend.SM100_FI_MXFP4_MXFP8_CUTLASS
-        ):
-            hidden_size = round_up(hidden_size, 128)
-        elif (
-            current_platform.is_rocm()
-            or current_mxfp4_backend == Mxfp4Backend.SM100_FI_MXFP4_MXFP8_TRTLLM
-            or current_mxfp4_backend == Mxfp4Backend.SM100_FI_MXFP4_BF16
-        ):
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py:268]: I just ran into a situation where the mxfp4 marlin kernels require 256 element padding.  Will this PR also address that or is it premature to remove this function?

`tests/entrypoints/openai/responses/test_harmony.py` fails if marlin is used for mxfp4.

See https://github.com/vllm-project/vllm/pull/32344/changes#diff-eddafffeb6f159f8c75f635d18a502fcfbf662a562b1ae7a8683a9790161a10b

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py:268]: Or maybe this needs to be incorporated into `maybe_roundup_layer_hidden_size`?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 3]
```diff
@@ -629,6 +575,19 @@ def _get_quant_method() -> FusedMoEMethodBase:
             moe_quant_params["intermediate_size_full"] = intermediate_size
 
         self.quant_method.create_weights(layer=self, **moe_quant_params)
+        # hidden_size may be padded in create_weights
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py]: Can you point to where this happens?

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
Comment (elvischenv) [reply vllm/model_executor/layers/fused_moe/layer.py:268]: If you look into the `create_weights()` inside `vllm/model_executor/layers/quantization/mxfp4.py`, you will see a completely duplicated logic with this function.

So the current padding logic inside FusedMoE init() will first call `maybe_roundup_hidden_size`, which includes a small part of padding logic for mxfp4. Then it will call `get_quant_method()` and `quant_method.create_weights()`. `create_weights()` will go through the whole padding logic again if it is using `Mxfp4MoEMethod`. cc @robertgshaw2-redhat

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 3 above)
Comment (elvischenv) [reply vllm/model_executor/layers/fused_moe/layer.py]: Answered in the previous comment.
The calling order is like `maybe_roundup_hidden_size` -> create MoE config `self.moe_config: FusedMoEConfig = FusedMoEConfig()` -> `get_quant_method()` -> `quant_method.create_weights()`.

There may be some paddings happening inside the `create_weights()` so need to update the MoE config.

The problem is where we should put the padding logic. Currently some are in `vllm/model_executor/layers/fused_moe/layer.py`, some are in `vllm/model_executor/layers/quantization/mxfp4.py`. And looks like PR #29008 depends on the logic in `layers/fused_moe/layer.py`.

cc @robertgshaw2-redhat

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py:268]: Can we keep all the logic here in `layer.py` instead of having it in two places?

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 3 above)
Comment (BowenBao) [reply vllm/model_executor/layers/fused_moe/layer.py]: fyi I have a (wip) draft to refactor the roundup logic at #34285, to move the kernel dependent rounding logic to quant_method. `fused_moe/layer.py` still needs to invoke it to update the sizes.

File: tests/compile/fusions_e2e/conftest.py (hunk 1 above)
Comment (nvpohanh) [reply tests/compile/fusions_e2e/conftest.py]: My original proposal would be make it part of the quantization config. for example: `--quantization mxfp4_mxfp8_act`
@elvischenv Could you open a GitHub issue so that we can discuss how to eliminate `VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8`? Thanks!

Comment (elvischenv) [reply tests/compile/fusions_e2e/conftest.py]: > Could you open a GitHub issue so that we can discuss how to eliminate VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8?

Created #34486

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 3 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py]: I think delegating the decision to the quant methods is the correct approach.

**Review by ProExpertProg — APPROVED** [review]

File: vllm/model_executor/layers/fused_moe/runner/default_moe_runner.py [hunk 4]
```diff
@@ -228,6 +228,12 @@ def __init__(
         self.batched_hidden_states: torch.Tensor | None = None
         self.batched_router_logits: torch.Tensor | None = None
 
+        # The padding in the forward pass can be skipped
+        self.skip_forward_padding = (
+            hasattr(self.quant_method, "support_skip_forward_padding")
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/runner/default_moe_runner.py]: Can you make this a method on `FusedMoEMethodBase` instead of an attribute?  It could default to `False`.

Comment (elvischenv) [reply vllm/model_executor/layers/fused_moe/runner/default_moe_runner.py]: Done

**Review by bnellnm — APPROVED** [review]

**Review by ProExpertProg — APPROVED** [review]
